chore(*): update Node.js from 8 to 12, update some dependencies

mgol opened the pull request:
# AngularJS is in LTS mode
We are no longer accepting changes that are not critical bug fixes into this project.
See https://blog.angular.io/stable-angularjs-and-long-term-support-7e077635ee9c for more detail.

<!-- General PR submission guidelines https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#submit-pr -->
**Does this PR fix a regression since 1.7.0, a security flaw, or a problem caused by a new browser version?**

<!-- If the answer is no, then we will not merge this PR -->


**What is the current behavior? (You can also link to an open issue here)**

Node.js 8 is used to build/test AngularJS.

**What is the new behavior (if this is a feature change)?**

Node.js 12 is used to build/test AngularJS.

**Does this PR introduce a breaking change?**

No.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
- ~~Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated~~
- ~~Fix/Feature: Tests have been added; existing tests pass~~

**Other information**:

Node.js 8 ends its support at the end of 2019. Node 12 will be supported
until April 2022 which is way past AngularJS end of support.

Some dependencies needed to be updated to make them work in Node.js 12.

TODO:
- [ ] Migrate to Gulp v4 😱

mgol committed:
chore(*): update Node.js from 8 to 12, update some dependencies

Node.js 8 ends its support at the end of 2019. Node 12 will be supported
until April 2022 which is way past AngularJS end of support.

Some dependencies needed to be updated to make them work in Node.js 12.

gkalpak reviewed (dismissed):
LGTM (as long as Travis is happy 😱)

gkalpak commented:
Rebased on master to see if c8a6e8450 would fix the failures.

Splaktar commented:
You can support Gulp `3.x` in NodeJS v12/14 using the approach that we used in AngularJS Material: https://github.com/angular/material/pull/11892

Upgrading to Gulp `v4` is generally considered too time intensive and not worth the effort for older projects that are close to ending LTS.

Splaktar commented on package.json:
```diff
@@ -9,8 +9,8 @@
-    "node": ">=8.12.0",
-    "yarn": ">=1.10.1",
+    "node": ">=12.14.1",
+    "yarn": ">=1.21.1",
```
Is there a specific reason why `yarn@1.19.0` or some other version isn't good enough? 

`1.10.1` is pretty old, but should we really drop support for ~`1.16`+ (for example that latest `yarn` available in macPorts in `1.19.1`)? 

Is there a specific feature in `1.21.1` that we now require?

mgol replied:
Yarn 1.21.1 contains a security fix, I'd rather people didn't use older versions: https://blog.npmjs.org/post/189618601100/binary-planting-with-the-npm-cli.

Splaktar replied:
OK, that makes sense.

Splaktar commented on package.json:
```diff
@@ -62,18 +62,18 @@
-    "karma-edge-launcher": "^0.4.2",
-    "karma-firefox-launcher": "^1.1.0",
-    "karma-ie-launcher": "^1.0.0",
+    "karma": "4.4.1",
```
I'm surprised that no test changes were required by this bump to Karma `v4`.

Splaktar reviewed (changes requested):
The primordials failures with Gulp `v3` are caused by NodeJS 12+ no longer accepting the `natives` override. The PR I linked above has another approach that solves these issues with `graceful-fs`.

mgol commented:
> The primordials failures with Gulp `v3` are caused by NodeJS 12+ no longer accepting the `natives` override. The PR I linked above has another approach that solves these issues with `graceful-fs`.

That's a great idea, thanks! I'll have a look later.

petebacondarwin committed:
chore(*): update protractor to latest version

petebacondarwin commented:
Adding a resolution for graceful-fs works at treat.

petebacondarwin commented:
@googlebot I consent

petebacondarwin committed:
chore(*): update saucelabs browsers

Splaktar commented:
Looks like there's a duplicate Googlebot here. We had this problem in AngularJS Material too. I think that there was one configured for the repo and one for the org. We were able to resolve it. I don't know if this affects only this PR or all PRs in the repo.

@josephperrott maybe dev-infra can resolve easily?

josephperrott commented:
@Splaktar I removed the duplicate Googlebot setup.  Should be good going forward.

petebacondarwin committed:
chore(ci): move over to CircleCi config

This commit gets rid of all references to Travis and, belatedly, Jenkins.
Now all CI is done on CircleCI and releases are run locally.

The CI no longer updates the docs and code.angularjs.org for jobs that are
not on the `master` branch.

During releases, the docs and code should be uploaded manually.

Splaktar commented on .circleci/config.yml:
```diff
@@ -0,0 +1,243 @@
+        type: string
+        default: medium
+    docker:
+      - image: circleci/node:12.14.1@sha256:f9de24fc0017059cc42ef7d07db060008af65a98b1f0cdd1ef3339213226bf6d
```
It looks like `12.16.3` is the latest. Any reason that you aren't using the `-browsers` builds? I.e. `circleci/node:12.16.3-browsers`.

Splaktar replied:
They come with Java and Firefox/Chrome and geckodriver/chromedriver installed: https://github.com/CircleCI-Public/circleci-dockerfiles/blob/master/node/images/12.16.3-stretch/browsers/Dockerfile

Splaktar commented on .circleci/config.yml:
```diff
@@ -0,0 +1,243 @@
+            # and network security service libraries (libnss3) & X11 Screen Saver extension library (libssx1)
+            # which are dependendies of chrome & needed for karma & protractor headless chrome tests.
+            # This is a very small install which takes around 7s in comparing to using the full
+            # circleci/node:x.x.x-browsers image.
```
Ah, this is why...

Splaktar commented on .circleci/env.sh:
```diff
@@ -0,0 +1,66 @@
+SL_EDGE,SL_EDGE-1"
+
+setPublicVar SAUCE_USERNAME "angular-framework";
+setSecretVar SAUCE_ACCESS_KEY "0c731274ed5f-cbc9-16f4-021a-9835e39f";
```
Should this be defined somewhere in CircleCI, outside of Git? Or is it safe to put here?

petebacondarwin replied:
It was copied from angular/angular project. But I guess we should hide it...

Splaktar commented on package.json:
```diff
@@ -102,7 +103,9 @@
     "//2": "(E.g. see https://github.com/gulpjs/gulp/issues/2162 and https://github.com/nodejs/node/issues/25132.)",
-    "natives": "1.1.6"
+    "natives": "1.1.6",
+    "//3": "`graceful-fs` needs to be pinned to support gulp 3",
```
```suggestion
    "//3": "`graceful-fs` needs to be pinned to support gulp 3 on NodeJS v12+",
```

Splaktar commented on protractor-travis-conf.js:
```diff
@@ -28,17 +28,18 @@ if (process.env.BROWSER_PROVIDER === 'browserstack') {
-      platform: 'OS X 10.11',
-      version: '9'
+      platform: 'OS X 10.14',
+      version: '12'
```
Is the goal for these to be using the `latest-1` version or the latest version?

petebacondarwin committed:
test(*): fix up some tests for Firefox

Splaktar commented on protractor-circleci-conf.js:
```diff
@@ -0,0 +1,44 @@
+  capabilitiesForSauceLabs({
+    browserName: 'chrome',
+    platform: 'OS X 10.14',
+    version: '81'
```
Chrome Stable just hit `83`.

petebacondarwin replied:
According to this https://saucelabs.com/platform/supported-browsers-devices SauceLabs only has Chrome 81 available right now.

Splaktar reviewed (approved):
LGTM. Just a note that Chrome M83 is the latest stable now and NodeJS 12.16.3 is the latest LTS.

gkalpak commented on .circleci/config.yml:
```diff
@@ -0,0 +1,380 @@
+    steps:
+      - custom_attach_workspace
+      - init_environment
+      - run: yarn -s grunt ci-checks
```
Why the `-s` here (but not elsewhere)?

petebacondarwin replied:
I copied it from the angular/angular project where it is used extensively. Then I forgot about it. I'll remove it because I don't think it makes much difference and is confusing if not used consistently.

gkalpak commented on .circleci/config.yml:
```diff
@@ -0,0 +1,380 @@
+          filters:
+            branches:
+              only:
+                - latest
```
Is `latest` a special value? Are there any docs on how it works?

petebacondarwin replied:
Previously deployments of the "latest" build, i.e. the releases that were tagged with a `latest` npm dist-tag also triggered deployment of the docs to a firebase hosting, and also a special version of the code to `snapshot-stable`.

This is more difficult in CircleCI, so I had this idea that we would simply keep a new `latest` git branch updated to the latest stable release, which would then trigger the deployments.
I never got it fleshed out properly - and now I am thinking, given the status of the project, that we can just run manual deployments when we run the release. WDYT?

gkalpak replied:
Why is it more difficult in CircleCI. It seems quite straight forward to me (just check `package.json > distTag` and abort the job if it is not `latest`).

petebacondarwin replied:
Hmm, I guess that could work. I was thinking of ways to avoid running the job in the first place.

gkalpak commented on scripts/travis/build.sh:
```diff
@@ -1,114 +0,0 @@
-    #   the commit is tagged
-    #   - or the branch is "master"
-    #   - or the branch distTag from package.json is "latest" (i.e. stable branch)
-    if [[ "$TRAVIS_TAG" != '' || "$TRAVIS_BRANCH" == master || "$DIST_TAG" == latest ]]; then
```
Has this logic been replicated in the new setup? I only see the corresponding job be run on  branches master and latest (whatever that means :grin:). What about tagged commits?

petebacondarwin replied:
This logic has not been mapped over 1 to 1. See https://github.com/angular/angular.js/pull/16915#discussion_r429653301

A tagged commit corresponds to a release, right? So if our release process includes deployment of docs and code, then we are done.  The only missing piece is the dist-tag updates...

gkalpak commented on scripts/travis/build.sh:
```diff
@@ -1,114 +0,0 @@
-    fi
-
-    if [[ "$DEPLOY_DOCS" == true || "$DEPLOY_CODE" == true ]]; then
-      grunt prepareDeploy
```
I see that `grunt prepareDeploy` is run for both `DEPLOY_CODE` and `DEPLOY_DOCS` here. In the new setup, it is only run in the `deploy-docs` job (but not the `deploy-code` one. Is that intended?

gkalpak reviewed (commented):
Post-merge review :grin:
Mostly some questions.

petebacondarwin commented on package.json:
```diff
@@ -98,11 +99,17 @@
   },
-  "dependencies": {},
+  "dependencies": {
+    "firebase": "^7.14.4",
```
Oops. I think that got included when I was messing about with trying out firebase deployments...

Splaktar replied:
Makes sense as a dev dependency 😄

petebacondarwin commented on package.json:
```diff
@@ -98,11 +99,17 @@
+  "dependencies": {
+    "firebase": "^7.14.4",
+    "firebase-tools": "^8.3.0",
+    "node-glob": "^1.2.0"
```
Similarly - I added this by accident.

petebacondarwin commented on scripts/release/release.sh:
```diff
@@ -36,7 +36,6 @@ function init {
 
```
Absolutely! I forgot to re-apply the `chmod`.

petebacondarwin commented on test/e2e/tests/anchor-scroll.spec.js:
```diff
@@ -123,6 +123,8 @@ describe('$anchorScroll', function() {
       var lastAnchor = anchors.last();
       var lastAnchorId = 'anchor-5';
 
+      if (browser.params.browser === 'firefox') return;
```
OK

soffieswan044 commented:
> # AngularJS is in LTS mode
> We are no longer accepting changes that are not critical bug fixes into this project.
> See https://blog.angular.io/stable-angularjs-and-long-term-support-7e077635ee9c for more detail.
> 
> **Does this PR fix a regression since 1.7.0, a security flaw, or a problem caused by a new browser version?**
> 
> **What is the current behavior? (You can also link to an open issue here)**
> 
> Node.js 8 is used to build/test AngularJS.
> 
> **What is the new behavior (if this is a feature change)?**
> 
> Node.js 12 is used to build/test AngularJS.
> 
> **Does this PR introduce a breaking change?**
> 
> No.
> 
> **Please check if the PR fulfills these requirements**
> 
> * [x]  The commit message follows our [guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#commits)
> * ~Fix/Feature: [Docs](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#documentation) have been added/updated~
> * ~Fix/Feature: Tests have been added; existing tests pass~
> 
> **Other information**:
> 
> Node.js 8 ends its support at the end of 2019. Node 12 will be supported
> until April 2022 which is way past AngularJS end of support.
> 
> Some dependencies needed to be updated to make them work in Node.js 12.
> 
> TODO:
> 
> * [ ]  Migrate to Gulp v4 😱
